### PR TITLE
feat: Create openapi specification

### LINKED ISSUE #18: Create OpenAPI specification
Generate OpenAPI spec from backend route stubs.

# Images
```bash
POST   /images # upload image (raw HTTP upload)
PATCH  /images/:id # set/modify image metadata, share/unshare, ...
GET    /images/:id?quality=(low|high)&limit=<number>&offset=<number>
GET    /images?quality=(low|high) # lists all images (with optional quality)
DELETE /images/:id
```
# Albums
```bash
GET    /albums?limit=<number>&offset=<number> # lists the user's albums (including shared ones)
POST   /albums # create a new album
PATCH  /albums/:id # update an album (title, ...), share/unshare
DELETE /albums/:id
POST   /albums/:id/images # add an image to an album
DELETE /albums/:id/images/:id # remove an image from an album
```
# Users
```bash
GET    /users/self # list own user's properties (name, email, ...)
PATCH  /users/self # change stuff about yourself
GET    /users # ADMIN ONLY
DELETE /users/:id # ADMIN ONLY
```
# Auth
Oauth ?

### PR DESCRIPTION
**Pull request by AndreasLrx**
Close #18

### CODE REVIEW

File: back/src/api/images.rs [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+};
+use crate::error_handler::APIError;
+use actix_multipart::Multipart;
+use actix_web::{delete, get, http, patch, post, web, Error, HttpResponse, Responder};
+use serde::{Deserialize, Serialize};
+use utoipa::ToSchema;
+
+const CONTEXT_PATH: &str = "/images";
+
+#[derive(Clone, Deserialize, Serialize, ToSchema)]
```
Comment (MisterPeModder) [inline back/src/api/images.rs:14]: use `#[serde(rename_all = "snake_case")]` to rename the fields.
Having upper case characters triggers me a bit.

File: back/src/api/mod.rs [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+#[allow(dead_code)]
+#[derive(Deserialize)]
+pub struct PaginationQuery {
+    limit: Option<i32>,
+    offset: Option<i32>,
+}
+
+pub fn query_payload_error_example() -> APIError {
+    APIError::query_payload_error(
+        "Query deserialize error: Unknown variant `lo`, expected one of `Low`, `Medium`, `High`",
```
Comment (MisterPeModder) [inline back/src/api/mod.rs]: Change this to snake_case if you can.

File: back/src/api/mod.rs [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    APIError::not_found_error("Image with id 15")
+}
+
+pub fn album_not_found_example() -> APIError {
+    APIError::not_found_error("Album with id 15")
+}
+
+#[derive(ToSchema)]
+#[schema(value_type = String, format = Binary)]
+pub struct Binary(String);
```
Comment (MisterPeModder) [inline back/src/api/mod.rs]: I get a dead code warning in nightly Rust.
Use this to fix it:
```rust
#[derive(ToSchema)]
#[schema(value_type = String, format = Binary)]
pub struct Binary(#[allow(dead_code)] String);
```

File: back/Cargo.toml [hunk 4]
```diff
@@ -34,6 +35,13 @@ jsonwebtoken = "^9.2.0"
 once_cell = "^1.19.0"
 futures-util = "~0.3.30"
 urlencoding = "^2.1.3"
+utoipa = { version = "4.2", features = ["actix_extras"] }
```
Comment (MisterPeModder) [inline back/Cargo.toml:38]: Use `^` for versions >=1.0 and `~` for versions under 1.0

File: back/src/api/albums.rs [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+use utoipa::ToSchema;
+
+const CONTEXT_PATH: &str = "/albums";
+
+#[derive(Deserialize, Serialize, ToSchema)]
+pub struct AlbumPost {
+    name: String,
+    tags: Vec<String>,
+    shared_with: Vec<String>,
+    images: Vec<u32>,
```
Comment (MisterPeModder) [inline back/src/api/albums.rs]: Ids are `Uuid`s

File: back/src/api/albums.rs [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+}
+
+#[derive(Deserialize, Serialize, ToSchema)]
+pub struct Album {
+    id: u32,
```
Comment (MisterPeModder) [inline back/src/api/albums.rs]: Ids are `Uuid`s

File: back/src/api/albums.rs [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+    albums: Vec<Album>,
+}
+
+/// Get an album by its id.
+#[allow(unused_variables)]
+#[allow(unreachable_code)]
+#[utoipa::path(
+    context_path = CONTEXT_PATH,
+    params(
+        ("id" = u16, Path, description="Identifier of the album", example=1)
```
Comment (MisterPeModder) [inline back/src/api/albums.rs]: Ids are `Uuid`s

File: back/src/api/albums.rs [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+    responses(
+        (status = StatusCode::OK, description = "Album retrieved successfully", body = Album, content_type = "application/json"),
+        (status = StatusCode::BAD_REQUEST, description = "Invalid path parameters", body = APIError, example = json!(path_error_example()), content_type = "application/json"),
+        (status = StatusCode::UNAUTHORIZED, description = "User not authenticated", body = APIError, example = json!(APIError::unauthorized_error()), content_type = "application/json"),
+        (status = StatusCode::NOT_FOUND, description = "Album not found (or user is forbidden to see it)", body = APIError, example = json!(album_not_found_example()), content_type = "application/json")
+    ),
+    tag="albums"
+)]
+#[get("/{id}")]
+pub async fn get_album(album_id: web::Path<u16>) -> impl Responder {
```
Comment (MisterPeModder) [inline back/src/api/albums.rs]: Ids are `Uuid`s

File: back/src/api/albums.rs [hunk 9]
```diff
@@ -0,0 +1,251 @@
...
+}
+
+/// Get a list of albums
+#[allow(unused_variables)]
+#[allow(unreachable_code)]
+#[utoipa::path(
+    context_path = CONTEXT_PATH,
+    params(
+        ("limit" = Option<u16>, Query, description="Number of albums to return", example=10),
+        ("offset" = Option<u16>, Query, description="Offset of the query in the albums list to return", example=0),
```
Comment (MisterPeModder) [inline back/src/api/albums.rs]: Is there a reason for `u16` offsets ?

File: back/src/api/albums.rs [hunk 10]
```diff
@@ -0,0 +1,251 @@
...
+    ),
+    tag="albums",
+    request_body(
+        description = "Album to edit",
+        content_type = "application/json",
+        content = AlbumPost
+    )
+)]
+#[patch("/{id}")]
+pub async fn edit_album(album_id: web::Path<u16>, patch: web::Json<AlbumPost>) -> impl Responder {
```
Comment (MisterPeModder) [inline back/src/api/albums.rs]: Ids are `Uuid`s

File: back/src/api/albums.rs [hunk 11]
```diff
@@ -0,0 +1,251 @@
...
+    HttpResponse::NoContent().finish()
+}
+
+/// Add an image to an album
+#[allow(unreachable_code)]
+#[allow(unused_variables)]
+#[utoipa::path(
+    context_path = CONTEXT_PATH,
+    params(
+        ("id" = u16, Path, description="Album to add the image to", example=1),
```
Comment (MisterPeModder) [inline back/src/api/albums.rs]: Ids are `Uuid`s

File: back/src/api/albums.rs [hunk 12]
```diff
@@ -0,0 +1,251 @@
...
+        (status = StatusCode::OK, description = "Successfully added", body = Album, content_type = "application/json"),
+        (status = StatusCode::BAD_REQUEST, description = "Invalid path parameters", body = APIError, example =json!(path_error_example()), content_type = "application/json"),
+        (status = StatusCode::UNAUTHORIZED, description = "User not authenticated", body = APIError, example = json!(APIError::unauthorized_error()), content_type = "application/json"),
+        (status = StatusCode::FORBIDDEN, description = "User has read only rights on the image/album (shared image/album)", body = APIError, example = json!(APIError::forbidden_error()), content_type = "application/json"),
+        (status = StatusCode::NOT_FOUND, description = "Album or image not found (or user is forbidden to see it)", body = APIError, example = json!(album_not_found_example()), content_type = "application/json")
+    ),
+    tag="albums"
+)]
+#[post("/{id}/images/{image_id}")]
+pub async fn add_image_to_album(path: web::Path<(u16, u16)>) -> impl Responder {
```
Comment (MisterPeModder) [inline back/src/api/albums.rs]: Ids are `Uuid`s

File: back/src/api/albums.rs [hunk 13]
```diff
@@ -0,0 +1,251 @@
...
+}
+
+/// Remove an image from an album
+#[allow(unreachable_code)]
+#[allow(unused_variables)]
+#[utoipa::path(
+    context_path = CONTEXT_PATH,
+    params(
+        ("id" = u16, Path, description="Album to remove the image from", example=1),
+        ("image_id" = u16, Path, description="Image to remove", example=1),
```
Comment (MisterPeModder) [inline back/src/api/albums.rs]: Ids are `Uuid`s

File: back/src/api/albums.rs [hunk 14]
```diff
@@ -0,0 +1,251 @@
...
+        (status = StatusCode::OK, description = "Successfully added", body = Album, content_type = "application/json"),
+        (status = StatusCode::BAD_REQUEST, description = "Invalid path parameters", body = APIError, example =json!(path_error_example()), content_type = "application/json"),
+        (status = StatusCode::UNAUTHORIZED, description = "User not authenticated", body = APIError, example = json!(APIError::unauthorized_error()), content_type = "application/json"),
+        (status = StatusCode::FORBIDDEN, description = "User has read only rights on the image/album (shared image/album)", body = APIError, example = json!(APIError::forbidden_error()), content_type = "application/json"),
+        (status = StatusCode::NOT_FOUND, description = "Album or image not found (or user is forbidden to see it)", body = APIError, example = json!(album_not_found_example()), content_type = "application/json")
+    ),
+    tag="albums"
+)]
+#[delete("/{id}/images/{image_id}")]
+pub async fn remove_image_from_album(path: web::Path<(u16, u16)>) -> impl Responder {
```
Comment (MisterPeModder) [inline back/src/api/albums.rs]: Ids are `Uuid`s

File: back/Cargo.toml
Comment (BaptistePgn) [inline back/Cargo.toml:1]: version fixe

File: back/src/api/images.rs [hunk 15]
```diff
@@ -0,0 +1,207 @@
...
+#[derive(Deserialize)]
+pub struct ImageQuery {
+    quality: Option<ImageQuality>,
+}
+
+#[allow(dead_code)]
+#[derive(Deserialize)]
+pub struct ImagesQuery {
+    limit: Option<i32>,
+    offset: Option<i32>,
```
Comment (MisterPeModder) [inline back/src/api/images.rs]: This is not consistent with the type in the spec (`u16`) although `i32`/`u32` is a better bit IMHO.

File: back/src/api/images.rs [hunk 16]
```diff
@@ -0,0 +1,207 @@
...
+#[derive(Deserialize, Serialize, ToSchema)]
+pub struct ImagePatch {
+    name: Option<String>,
+    tags: Option<Vec<String>>,
+    shared_with: Option<Vec<String>>,
+}
+
+#[derive(Deserialize, Serialize, ToSchema)]
+pub struct ImageMetaData {
+    id: u32,
```
Comment (MisterPeModder) [inline back/src/api/images.rs]: Ids are `Uuid`s

File: back/src/api/images.rs [hunk 17]
```diff
@@ -0,0 +1,207 @@
...
+}
+
+#[derive(Deserialize, Serialize, ToSchema)]
+pub struct ImagesMetaData {
+    images: Vec<ImageMetaData>,
+}
+
+#[derive(Serialize, Deserialize, ToSchema)]
+pub struct ImageUploadResponse {
+    id: u32,
```
Comment (MisterPeModder) [inline back/src/api/images.rs]: Ids are `Uuid`s

File: back/src/api/images.rs [hunk 18]
```diff
@@ -0,0 +1,207 @@
...
+}
+
+/// Get an image by its id.
+#[allow(unused_variables)]
+#[allow(unreachable_code)]
+#[utoipa::path(
+    context_path = CONTEXT_PATH,
+    params(
+        ("id" = u16, Path, description="Identifier of the image", example=1),
```
Comment (MisterPeModder) [inline back/src/api/images.rs]: Ids are `Uuid`s

File: back/src/api/images.rs [hunk 19]
```diff
@@ -0,0 +1,207 @@
...
+            ("Invalid path parameters" = (value = json!(path_error_example())))),
+            content_type = "application/json"
+        ),
+        (status = StatusCode::UNAUTHORIZED, description = "User not authenticated", body = APIError, example = json!(APIError::unauthorized_error()), content_type = "application/json"),
+        (status = StatusCode::NOT_FOUND, description = "Image not found (or user is forbidden to see it)", body = APIError, example = json!(image_not_found_example()), content_type = "application/json")
+    ),
+    tag="images"
+)]
+#[get("/{id}")]
+pub async fn get_image(img_id: web::Path<u16>, query: web::Query<ImageQuery>) -> impl Responder {
```
Comment (MisterPeModder) [inline back/src/api/images.rs]: Ids are `Uuid`s

File: back/src/api/images.rs [hunk 20]
```diff
@@ -0,0 +1,207 @@
...
+///
+/// This method returns the metadata of the images not the effective images. The client must make a request for each image independently.
+/// The list can be filtered by quality, and paginated.
+#[allow(unused_variables)]
+#[allow(unreachable_code)]
+#[utoipa::path(
+    context_path = CONTEXT_PATH,
+    params(
+        ("limit" = Option<u16>, Query, description="Number of images to return", example=10),
+        ("offset" = Option<u16>, Query, description="Offset of the query in the image list to return", example=0),
```
Comment (MisterPeModder) [inline back/src/api/images.rs]: Ids are `Uuid`s

File: back/src/api/images.rs [hunk 21]
```diff
@@ -0,0 +1,207 @@
...
+    HttpResponse::Ok().json(ImageUploadResponse { id: 1 })
+}
+
+/// Set/modfiy image metadata, share/unshare, ...
+#[allow(unreachable_code)]
+#[allow(unused_variables)]
+#[utoipa::path(
+    context_path = CONTEXT_PATH,
+    params(
+        ("id" = u16, Path, description="Image to edit", example=1),
```
Comment (MisterPeModder) [inline back/src/api/images.rs]: Ids are `Uuid`s

File: back/src/error_handler.rs [hunk 22]
```diff
@@ -0,0 +1,191 @@
...
+    HttpResponse, HttpResponseBuilder, Result,
+};
+use actix_web::{HttpRequest, ResponseError};
+use futures::executor::block_on;
+use serde::{Deserialize, Serialize};
+use std::str::from_utf8;
+use utoipa::ToSchema;
+
+#[derive(Clone, Deserialize, Serialize, Debug, ToSchema)]
+pub enum ApiErrorCode {
```
Comment (MisterPeModder) [inline back/src/error_handler.rs:15]: Not necessary, but I put `#[serde(rename_all = "SREAMING_SNAKE_CASE")]` on my error codes to have them apprear like `"code": "INVALID_CREDENTIALS"`, makes them look more important

**Review by MisterPeModder — CHANGES_REQUESTED** [review]

File: back/src/error_handler.rs
Comment (BaptistePgn) [inline back/src/error_handler.rs:1]: APIError attention majuscule "API"

**Review by MisterPeModder — APPROVED** [review]

**Pull request merged**